File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/query/SubQueryTest.java
package org.babyfish.jimmer.sql.query;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.Expression;
import org.babyfish.jimmer.sql.ast.Predicate;
import org.babyfish.jimmer.sql.ast.query.TypedSubQuery;
import org.babyfish.jimmer.sql.common.AbstractQueryTest;
import org.babyfish.jimmer.sql.model.*;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.Arrays;

public class SubQueryTest extends AbstractQueryTest {

    @Test
    public void testColumnInSubQuery() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                    q.where(
                            book.id().in(
                                    getLambdaClient().createSubQuery(q, AuthorTableEx.class, (sq, author) -> {
                                        sq.where(author.firstName().eq("Alex"));
                                        return sq.select(author.books().id());
                                    })
                            )
                    );
                    return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "where tb_1_.ID in (" +
                                        "select tb_3_.BOOK_ID " +
                                        "from AUTHOR tb_2_ " +
                                        "inner join BOOK_AUTHOR_MAPPING tb_3_ on tb_2_.ID = tb_3_.AUTHOR_ID " +
                                        "where tb_2_.FIRST_NAME = ?" +
                                    ")"
                    );
                    ctx.variables("Alex");
                }
        );
    }

    @Test
    public void testTwoColumnsInSubQuery() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                    q.where(
                        Expression.tuple(book.name(), book.price()).in(
                                getLambdaClient().createSubQuery(q, BookTableEx.class, (sq, book2) ->
                                        sq
                                                .groupBy(book2.name())
                                                .select(
                                                        book2.name(),
                                                        book2.price().max()
                                                )
                                )
                        )
                    );
                    return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "where (tb_1_.NAME, tb_1_.PRICE) in (" +
                                        "select tb_2_.NAME, max(tb_2_.PRICE) " +
                                        "from BOOK tb_2_ " +
                                        "group by tb_2_.NAME" +
                                    ")"
                    );
                }
        );
    }

    @Test
    public void testExists() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                        q.where(
                                getLambdaClient().createWildSubQuery(q, AuthorTableEx.class, (sq, author) -> {
                                    sq.where(
                                            book.eq(author.books()),
                                            author.firstName().eq("Alex")
                                    );
                                }).exists()
                        );
                        return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ where exists(" +
                                        "select 1 " +
                                        "from AUTHOR tb_2_ " +
                                        "inner join BOOK_AUTHOR_MAPPING tb_3_ on tb_2_.ID = tb_3_.AUTHOR_ID " +
                                        "where tb_1_.ID = tb_3_.BOOK_ID " +
                                        "and tb_2_.FIRST_NAME = ?" +
                                    ")"
                    );
                    ctx.variables("Alex");
                }
        );
    }

    @Test
    public void testExistsWithTypedQuery() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                    q.where(
                            getLambdaClient().createSubQuery(q, AuthorTableEx.class, (sq, author) -> {
                                sq.where(
                                        book.eq(author.books()),
                                        author.firstName().eq("Alex")
                                );
                                return sq.select(author);
                            }).exists()
                    );
                    return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ where exists(" +
                                    "select 1 " +
                                    "from AUTHOR tb_2_ " +
                                    "inner join BOOK_AUTHOR_MAPPING tb_3_ on tb_2_.ID = tb_3_.AUTHOR_ID " +
                                    "where tb_1_.ID = tb_3_.BOOK_ID " +
                                    "and tb_2_.FIRST_NAME = ?" +
                                    ")"
                    );
                    ctx.variables("Alex");
                }
        );
    }

    @Test
    public void testSubQueryAsSimpleExpression() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                    q.where(
                            book.price().gt(
                                    getLambdaClient().createSubQuery(q, BookTableEx.class, (sq, book2) -> {
                                        return sq.select(book2.price().avgAsDecimal().coalesce(BigDecimal.ZERO));
                                    })
                            )
                    );
                    return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ where tb_1_.PRICE > (" +
                                        "select coalesce(avg(tb_2_.PRICE), ?) from BOOK tb_2_" +
                                    ")"
                    );
                    ctx.variables(BigDecimal.ZERO);
                }
        );
    }

    @Test
    public void testSubQueryAsSelectionOrderByClause() {
        executeAndExpect(
                getLambdaClient().createQuery(BookStoreTable.class, (q, store) -> {
                    TypedSubQuery<BigDecimal> subQuery =
                            getLambdaClient().createSubQuery(q, BookTableEx.class, (sq, book) -> {
                                sq.where(store.eq(book.store()));
                                return sq.select(
                                        book.price().avgAsDecimal().coalesce(BigDecimal.ZERO)
                                );
                            });
                    q.orderBy(subQuery.desc());
                    return q.select(store, subQuery);
                }),
                ctx -> {
                    ctx.sql(
                            "select " +
                                        "tb_1_.ID, " +
                                        "tb_1_.NAME, " +
                                        "tb_1_.WEBSITE, " +
                                        "tb_1_.VERSION, " +
                                        "(" +
                                            "select coalesce(avg(tb_2_.PRICE), ?) " +
                                            "from BOOK tb_2_ " +
                                            "where tb_1_.ID = tb_2_.STORE_ID" +
                                        ") " +
                                    "from BOOK_STORE tb_1_ " +
                                    "order by (" +
                                        "select coalesce(avg(tb_2_.PRICE), ?) " +
                                        "from BOOK tb_2_ " +
                                        "where tb_1_.ID = tb_2_.STORE_ID" +
                                    ") desc"
                    );
                    ctx.variables(BigDecimal.ZERO, BigDecimal.ZERO);
                }
        );
    }

    @Test
    public void testSubQueryWithAny() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                    q.where(
                            book.id().eq(
                                    getLambdaClient().createSubQuery(q, AuthorTableEx.class, (sq, author) -> {
                                        sq.where(author.firstName().in(Arrays.asList("Alex", "Bill")));
                                        return sq.select(author.books().id());
                                    }).any()
                            )
                    );
                    return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "where tb_1_.ID = any(" +
                                        "select tb_3_.BOOK_ID " +
                                        "from AUTHOR tb_2_ " +
                                        "inner join BOOK_AUTHOR_MAPPING tb_3_ on tb_2_.ID = tb_3_.AUTHOR_ID " +
                                        "where tb_2_.FIRST_NAME in (?, ?)" +
                                    ")"
                    );
                    ctx.variables("Alex", "Bill");
                }
        );
    }

    @Test
    public void testSubQueryWithAll() {
        executeAndExpect(
                getLambdaClient().createQuery(BookTable.class, (q, book) -> {
                    q.where(
                            book.id().eq(
                                    getLambdaClient().createSubQuery(q, AuthorTableEx.class, (sq, author) -> {
                                        sq.where(author.firstName().in(Arrays.asList("Alex", "Bill")));
                                        return sq.select(author.books().id());
                                    }).all()
                            )
                    );
                    return q.select(book);
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "where tb_1_.ID = all(" +
                                        "select tb_3_.BOOK_ID " +
                                        "from AUTHOR tb_2_ " +
                                        "inner join BOOK_AUTHOR_MAPPING tb_3_ on tb_2_.ID = tb_3_.AUTHOR_ID " +
                                        "where tb_2_.FIRST_NAME in (?, ?)" +
                                    ")"
                    );
                    ctx.variables("Alex", "Bill");
                }
        );
    }

    @Test
    public void testSelectCount() {
        BookTable table = BookTable.$;
        AuthorTableEx author = AuthorTableEx.$;
        JSqlClient sqlClient = getSqlClient();
        executeAndExpect(
                sqlClient.createQuery(table)
                        .where(
                                sqlClient.createSubQuery(author)
                                        .where(author.books().eq(table))
                                        .selectCount()
                                        .gt(1L)
                        )
                        .select(table),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "where (" +
                                    "--->select count(1) " +
                                    "--->from AUTHOR tb_2_ " +
                                    "--->inner join BOOK_AUTHOR_MAPPING tb_3_ on tb_2_.ID = tb_3_.AUTHOR_ID " +
                                    "--->where tb_3_.BOOK_ID = tb_1_.ID" +
                                    ") > ?"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"id\":\"e110c564-23cc-4811-9e81-d587a13db634\"," +
                                    "--->--->\"name\":\"Learning GraphQL\"," +
                                    "--->--->\"edition\":1," +
                                    "--->--->\"price\":50.00," +
                                    "--->--->\"storeId\":\"d38c10da-6be8-4924-b9b9-5e81899612a0\"" +
                                    "--->},{" +
                                    "--->--->\"id\":\"b649b11b-1161-4ad2-b261-af0112fdd7c8\"," +
                                    "--->--->\"name\":\"Learning GraphQL\"," +
                                    "--->--->\"edition\":2," +
                                    "--->--->\"price\":55.00," +
                                    "--->--->\"storeId\":\"d38c10da-6be8-4924-b9b9-5e81899612a0\"" +
                                    "--->},{" +
                                    "--->--->\"id\":\"64873631-5d82-4bae-8eb8-72dd955bfc56\"," +
                                    "--->--->\"name\":\"Learning GraphQL\"," +
                                    "--->--->\"edition\":3," +
                                    "--->--->\"price\":51.00," +
                                    "--->--->\"storeId\":\"d38c10da-6be8-4924-b9b9-5e81899612a0\"" +
                                    "}]"
                    );
                }
        );
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/query/SubQueryTest.kt
package org.babyfish.jimmer.sql.kt.query

import org.babyfish.jimmer.sql.ast.LikeMode
import org.babyfish.jimmer.sql.kt.ast.expression.*
import org.babyfish.jimmer.sql.kt.common.AbstractQueryTest
import org.babyfish.jimmer.sql.kt.model.*
import org.babyfish.jimmer.sql.kt.model.classic.book.Book
import org.babyfish.jimmer.sql.kt.model.classic.book.edition
import org.babyfish.jimmer.sql.kt.model.classic.book.name
import org.babyfish.jimmer.sql.kt.model.classic.book.store
import org.babyfish.jimmer.sql.kt.model.classic.store.BookStore
import org.babyfish.jimmer.sql.kt.model.classic.store.id
import org.babyfish.jimmer.sql.kt.model.classic.store.name
import kotlin.test.Test

class SubQueryTest : AbstractQueryTest() {

    @Test
    fun testAsExpression() {
        executeAndExpect(
            sqlClient.createQuery(BookStore::class) {
                val childCount = subQuery(Book::class) {
                    where(table.store eq parentTable)
                    select(count(table))
                }
                orderBy(childCount)
                select(table, childCount)
            }
        ) {
            sql(
                """select 
                    |--->tb_1_.ID, tb_1_.NAME, tb_1_.VERSION, tb_1_.WEBSITE, 
                    |--->(
                    |--->--->select count(tb_2_.ID) 
                    |--->--->from BOOK tb_2_ 
                    |--->--->where 
                    |--->--->--->tb_2_.STORE_ID = tb_1_.ID
                    |--->) 
                    |from BOOK_STORE tb_1_ 
                    |order by (
                    |--->select count(tb_2_.ID) 
                    |--->from BOOK tb_2_ 
                    |--->where 
                    |--->--->tb_2_.STORE_ID = tb_1_.ID
                    |) asc""".trimMargin()
            )
            rows {
                contentEquals(
                    """[
                        |--->Tuple2(
                        |--->--->_1={
                        |--->--->--->"id":2,
                        |--->--->--->"name":"MANNING",
                        |--->--->--->"version":0,
                        |--->--->--->"website":null
                        |--->--->}, 
                        |--->--->_2=3
                        |--->), 
                        |--->Tuple2(
                        |--->--->_1={
                        |--->--->--->"id":1,
                        |--->--->--->"name":"O'REILLY",
                        |--->--->--->"version":0,
                        |--->--->--->"website":null
                        |--->--->}, 
                        |--->--->_2=9
                        |--->)
                        |]""".trimMargin(),
                    it.toString()
                )
            }
        }
    }

    @Test
    fun testSingleColumnIn() {
        executeAndExpect(
            sqlClient.createQuery(Book::class) {
                where(
                    table.store.id valueIn subQuery(BookStore::class) {
                        where(table.name.like("MAN", LikeMode.START))
                        select(table.id)
                    }
                )
                select(table)
            }
        ) {
            sql(
                """select 
                    |tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID 
                    |from BOOK tb_1_ 
                    |where tb_1_.STORE_ID in (
                    |--->select tb_3_.ID 
                    |--->from BOOK_STORE tb_3_ 
                    |--->where tb_3_.NAME like ?
                    |)""".trimMargin()
            )
            variables("MAN%")
            rows {
                contentEquals(
                    """[
                        |--->{
                        |--->--->"id":10,
                        |--->--->"name":"GraphQL in Action",
                        |--->--->"edition":1,
                        |--->--->"price":80.00,
                        |--->--->"store":{"id":2}
                        |--->}, {
                        |--->--->"id":11,
                        |--->--->"name":"GraphQL in Action",
                        |--->--->"edition":2,
                        |--->--->"price":81.00,
                        |--->--->"store":{"id":2}
                        |--->}, {
                        |--->--->"id":12,
                        |--->--->"name":"GraphQL in Action",
                        |--->--->"edition":3,
                        |--->--->"price":80.00,
                        |--->--->"store":{"id":2}
                        |--->}
                        |]""".trimMargin(),
                    it.toString()
                )
            }
        }
    }

    @Test
    fun testMultipleColumnIn() {
        executeAndExpect(
            sqlClient.createQuery(Book::class) {
                where(tuple(table.name, table.edition) valueIn subQuery(Book::class) {
                    groupBy(table.name)
                    select(table.name, max(table.edition).asNonNull())
                })
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID 
                    |from BOOK tb_1_ 
                    |where (tb_1_.NAME, tb_1_.EDITION) in (
                    |--->select tb_2_.NAME, max(tb_2_.EDITION) 
                    |--->from BOOK tb_2_ 
                    |--->group by tb_2_.NAME
                    |)""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"id":3,
                    |--->--->"name":"Learning GraphQL",
                    |--->--->"edition":3,
                    |--->--->"price":51.00,
                    |--->--->"store":{"id":1}
                    |--->},{
                    |--->--->"id":6,
                    |--->--->"name":"Effective TypeScript",
                    |--->--->"edition":3,
                    |--->--->"price":88.00,
                    |--->--->"store":{"id":1}
                    |--->},{
                    |--->--->"id":9,
                    |--->--->"name":"Programming TypeScript",
                    |--->--->"edition":3,
                    |--->--->"price":48.00,
                    |--->--->"store":{"id":1}
                    |--->},{
                    |--->--->"id":12,
                    |--->--->"name":"GraphQL in Action",
                    |--->--->"edition":3,
                    |--->--->"price":80.00,
                    |--->--->"store":{"id":2}
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testAny() {
        executeAndExpect(
            sqlClient.createQuery(Book::class) {
                where(
                    tuple(table.name, table.edition) eq any(
                        subQuery(Book::class) {
                            groupBy(table.name)
                            select(
                                table.name,
                                max(table.edition).asNonNull()
                            )
                        }
                    )
                )
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID 
                    |from BOOK tb_1_ 
                    |where (tb_1_.NAME, tb_1_.EDITION) = any(
                    |--->select tb_2_.NAME, max(tb_2_.EDITION) 
                    |--->from BOOK tb_2_ 
                    |--->group by tb_2_.NAME
                    |)""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"id":3,
                    |--->--->"name":"Learning GraphQL",
                    |--->--->"edition":3,
                    |--->--->"price":51.00,
                    |--->--->"store":{"id":1}
                    |--->},{
                    |--->--->"id":6,
                    |--->--->"name":"Effective TypeScript",
                    |--->--->"edition":3,
                    |--->--->"price":88.00,
                    |--->--->"store":{"id":1}
                    |--->},{
                    |--->--->"id":9,
                    |--->--->"name":"Programming TypeScript",
                    |--->--->"edition":3,
                    |--->--->"price":48.00,
                    |--->--->"store":{"id":1}
                    |--->},{
                    |--->--->"id":12,
                    |--->--->"name":"GraphQL in Action",
                    |--->--->"edition":3,
                    |--->--->"price":80.00,
                    |--->--->"store":{"id":2}
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testExists() {
        executeAndExpect(
            sqlClient.createQuery(BookStore::class) {
                where(
                    notExists(
                        subQuery(Book::class) {
                            where(
                                table.store eq parentTable,
                                or(
                                    table.name ilike "type",
                                    table.name ilike "script"
                                )
                            )
                            select(constant(1))
                        }
                    )
                )
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.VERSION, tb_1_.WEBSITE 
                    |from BOOK_STORE tb_1_ 
                    |where not exists(
                    |--->select 1 
                    |--->from BOOK tb_2_ 
                    |--->where 
                    |--->--->tb_2_.STORE_ID = tb_1_.ID 
                    |--->and (
                    |--->--->--->lower(tb_2_.NAME) like ? 
                    |--->--->or 
                    |--->--->--->lower(tb_2_.NAME) like ?
                    |--->)
                    |)""".trimMargin()
            )
            variables("%type%", "%script%")
            rows(
                """[{"id":2,"name":"MANNING","version":0,"website":null}]"""
            )
        }
    }

    @Test
    fun testExistsOnWildSubQuery() {
        executeAndExpect(
            sqlClient.createQuery(BookStore::class) {
                where(
                    notExists(
                        wildSubQuery(Book::class) {
                            where(
                                table.store eq parentTable,
                                or(
                                    table.name ilike "type",
                                    table.name ilike "script"
                                )
                            )
                        }
                    )
                )
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.VERSION, tb_1_.WEBSITE 
                    |from BOOK_STORE tb_1_ 
                    |where not exists(
                    |--->select 1 
                    |--->from BOOK tb_2_ 
                    |--->where 
                    |--->--->tb_2_.STORE_ID = tb_1_.ID 
                    |--->and (
                    |--->--->--->lower(tb_2_.NAME) like ? 
                    |--->--->or 
                    |--->--->--->lower(tb_2_.NAME) like ?
                    |--->)
                    |)""".trimMargin()
            )
            variables("%type%", "%script%")
            rows(
                """[{"id":2,"name":"MANNING","version":0,"website":null}]"""
            )
        }
    }
}